src/otx/engine/engine.py

```diff
@@ -0,0 +1,486 @@
+            **kwargs: Additional keyword arguments for pl.Trainer configuration.
+
+        Returns:
+            dict: Dictionary containing the callback metrics from the trainer.
+
+        Example:
+        >>> engine.test(
+            datamodule=OTXDataModule(),
+            checkpoint="checkpoint.ckpt",
+        )
+
+        CLI Usage:
+            1. you can pick a model.
+                ```python
+                otx test
+                    --model <CONFIG | CLASS_PATH_OR_NAME> --data_root <DATASET_PATH, str>
+                    --checkpoint <CKPT_PATH, str>
+                ```
+            2. If you have a ready configuration file, run it like this.
+                ```python
+                otx test --config <CONFIG_PATH, str> --checkpoint <CKPT_PATH, str>
+                ```
+        """
+        lit_module = self._build_lightning_module(
+            model=self.model,
+            optimizer=self.optimizer,
+            scheduler=self.scheduler,
+        )
+        if datamodule is None:
+            datamodule = self.datamodule
+        lit_module.meta_info = datamodule.meta_info
+
+        self._build_trainer(**kwargs)
+
+        self.trainer.test(
+            model=lit_module,
+            dataloaders=datamodule,
+            ckpt_path=str(checkpoint) if checkpoint is not None else self.checkpoint,
+        )
+
+        return self.trainer.callback_metrics
+
+    def predict(
+        self,
+        checkpoint: str | Path | None = None,
+        datamodule: EVAL_DATALOADERS | OTXDataModule | None = None,
+        return_predictions: bool | None = None,
+        **kwargs,
+    ) -> list | None:
+        """Run predictions using the specified model and data.
+
+        Args:
+            datamodule (EVAL_DATALOADERS | OTXDataModule | None, optional): The data module to use for predictions.
+            checkpoint (str | Path | None, optional): The path to the checkpoint file to load the model from.
+            return_predictions (bool | None, optional): Whether to return the predictions or not.
+            **kwargs: Additional keyword arguments for pl.Trainer configuration.
+
+        Returns:
+            list | None: The predictions if `return_predictions` is True, otherwise None.
+
+        Example:
+        >>> engine.predict(
+            datamodule=OTXDataModule(),
+            checkpoint="checkpoint.ckpt",
+            return_predictions=True,
+        )
+
+        CLI Usage:
+            1. you can pick a model.
+                ```python
+                otx predict
+                    --model <CONFIG | CLASS_PATH_OR_NAME> --data_root <DATASET_PATH, str>
+                    --checkpoint <CKPT_PATH, str>
+                ```
+            2. If you have a ready configuration file, run it like this.
+                ```python
+                otx predict --config <CONFIG_PATH, str> --checkpoint <CKPT_PATH, str>
+                ```
+        """
+        lit_module = self._build_lightning_module(
```

Comment:
Since we build a lightning module in each method we can possibly support passing the model in each method as well.
predict/train/test/... can have model as an optional argument. Since the model setter already supports OTX model or a string we can do something like this.
```python
def predict(self,
    model: OTXModel | str | None = None,
    ...
):
    if model:
        self.update_model(model)
        or
        self.model = model
```

One thing I can see is that the `get_model` method also takes in `num_classes` as a parameter. But I am sure we can design a system around this. Something like

```python
def update_model(model: OTXModel | str, datamodule: Datamodule | None = None):
    num_classes = datamodule.num_classes if datamodule else self.datamodule.num_classes
    self._auto_configurator.get_model(
                model=model,
                num_classes=num_classes,
     )
```

Unless I am missing something and model is required in `__init__` for some other initializations as well.

Reply:
This is something to think about a bit more. For one thing, it can be confusing for users when models are overridden (when and where), and it's pretty hard to modify the num_classes of an already built model.
Instead we could consider a setter function, which probably already exists in engine.py in some form.
What do you think?